### PR TITLE
15.ai vox: 20 year old spaceman game, cutting edge AI technology

### PR DESCRIPTION
**Pull request by Iamgoofball**
[FEATURE]
## About the PR 

AI players can now broadcast globally with cutting edge neural network voice synthesis technology powered by 15.ai.

Check it out at http://15.ai

dreamseeker_EGt6hkgG8E

More characters coming soon.

## Why's this needed? 

AIs will appreciate being able to vocalize their thoughts to the crew in various fun voices.

## Changelog
```
(u)Iamgoofball and 15.ai:
(*)Using the power of 15.ai, the AI can now vocalize their thoughts with a variety of voices.
```
https://file.house/qw2S.webm
https://file.house/vIWd.webm

### CODE REVIEW

File: code/WorkInProgress/fifteenAI_vox.dm [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	if(announcing_vox > world.time)
+		boutput(src, "<span class='notice'>The VOX system is still recharging. Please wait about 30 seconds past your last usage of it.</span>")
+		return
+	var/character_to_use = input(src, "Choose what 15.ai character to use:", "15.ai Character Choice")  as null|anything in friendly_voice_names
+	if(!character_to_use)
+		return
+	var/emotion_to_use = input(src, "Choose what emotion to use! Higher numbers are more emotive:", "Emotion Choice")  as null|anything in available_emotions[friendly_voice_names[character_to_use]]
+	if(!emotion_to_use)
+		return
+	var/max_characters = 300 // magic number but its the cap 15 allows
```
Comment (ZeWaka) [inline code/WorkInProgress/fifteenAI_vox.dm:119]:
```suggestion
	var/const/max_characters = 300 // magic number but its the cap 15 allows
```

File: code/WorkInProgress/fifteenAI_vox.dm [hunk 2]
```diff
@@ -0,0 +1,166 @@
+//http request stuff stolen from /tg/ to go along with the rust-g http request stuff, modified by goof for 15.ai stuff
+/datum/http_request
```
Comment (ZeWaka) [inline code/WorkInProgress/fifteenAI_vox.dm:2]: this should be in a different file

File: code/WorkInProgress/fifteenAI_vox.dm [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+/datum/http_response
+	var/status_code
+	var/body
+	var/list/headers
+
+	var/errored = FALSE
+	var/error
+
+#define UNTIL(X) while(!(X)) sleep(world.tick_lag)
```
Comment (ZeWaka) [inline code/WorkInProgress/fifteenAI_vox.dm:80]: belongs in _macros.dm

File: code/WorkInProgress/fifteenAI_vox.dm [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	last_announcement = message
+
+	announcing_vox = world.time + 30 SECONDS
+
+	message_admins("[key_name(src)] started making a 15.AI announcement with the following message: [message]")
+	play_vox_word(message, friendly_voice_names[character_to_use], emotion_to_use, src, src.z, null)
+
+
+/proc/play_vox_word(message, character, emotion, mob/living/silicon/ai/speaker, z_level, mob/only_listener)
+	var/api_url = "https://api.fifteen.ai/app/getAudioFile"
```
Comment (ZeWaka) [inline code/WorkInProgress/fifteenAI_vox.dm:142]:
```suggestion
	var/const/api_url = "https://api.fifteen.ai/app/getAudioFile"
```

File: config/admins.txt [hunk 5]
```diff
@@ -2,4 +2,5 @@
 #The available ranks can be found in /proc/rank_to_level().
 #Most of the time, you'll want to make yourself a Host.
 #To do so, type your BYOND key with no spaces, capitals or other punctuation (including underscores!), a dash with spaces on both side, then the rank you want.
-#For example, the key "ZeWaka" would be given Host rank like this: zewaka - Host
\ No newline at end of file
+#For example, the key "ZeWaka" would be given Host rank like this: zewaka - Host
+iamgoofball - Host
```
Comment (ZeWaka) [inline config/admins.txt:6]:
```suggestion
```

File: code/WorkInProgress/fifteenAI_vox.dm [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	req.prepare(RUSTG_HTTP_METHOD_POST, api_url, "{\"character\":\"[character]\",\"text\":\"[message]\",\"emotion\":\"[emotion]\"}", list("Content-Type" = "application/json", "User-Agent" = "Goonstation server"), json_encode(list("output_filename" = "data/vox_[vox_voice_number].wav")))
+	req.begin_async()
+	UNTIL(req.is_complete())
+	var/datum/http_response/res = req.into_response()
+	if(res.status_code == 200)
+		message_admins("[key_name(speaker)] finished making a 15.AI announcement with the following message: [message]")
+		speaker.say(";[message]")
+		for(var/C in clients)
+			if(C)
+				C << sound("data/vox_[vox_voice_number].wav", environment = -1)
```
Comment (ZeWaka) [inline code/WorkInProgress/fifteenAI_vox.dm:155]:
```suggestion
				UNLINT(C << sound("data/vox_[vox_voice_number].wav", environment = -1))
```

**Comment by Kyle2143** [comment]
Afaik we currently do not let AIs or any non-admins use dectalk or Vox, so I don't know why we'd give them some other speech synthesizer. Fine to add in for admin nonsense, but such a change to AI will require admin discussion first.

**Comment by Iamgoofball** [comment]
something is extremely wrong with _std/code/_macros.dm

whatever that did, it made it conflict on here and it refuses to fix itself when pulling from master locally

no conflicting files locally, git just fails silently

**Comment by Iamgoofball** [comment]
TortoiseGitProc_F488E1S2ME
reverting that macros change for now

**Comment by LanceFyre99** [comment]
I agree that this shouldn't just be handed out by default, but I desperately want this implemented as something that can be granted in a manner similar to vox. It's way too funny.

**Comment by Iamgoofball** [comment]
I fundamentally disagree that this will be an issue being available by default.

Players who spam it can be dealt with by administrators, and the system limitations prevent obnoxious abuse of it. There's almost certainly going to be an "ooh shiny" effect for the first week or so, I don't doubt that, but I really don't think users will spam it nearly as much as people expect it to be spammed.

**Comment by ZeWaka** [comment]
> 
> 
> I fundamentally disagree that this will be an issue being available by default.
> 
> Players who spam it can be dealt with by administrators, and the system limitations prevent obnoxious abuse of it. There's almost certainly going to be an "ooh shiny" effect for the first week or so, I don't doubt that, but I really don't think users will spam it nearly as much as people expect it to be spammed.

After hearing people spam our radio station record player for perhaps a year now, I can tell you that people will definitely overuse this.

**Comment by Iamgoofball** [comment]
Then simply hit those players with a warning for spam, much like anything else in the game. Cooldowns can be adjusted to reduce spam potential, admins are informed when people try to use voices as well. I don't think it'll be nearly as big an issue as theorized.

At the end of the day, /tg/'s playerbase of dumb fucking children with zero impulse control behaved with this tool.

If /tg/'s dumb fucking children with zero impulse control players can handle it, the grown ass adults at goon definitely can handle it.

**Comment by Iamgoofball** [comment]
The ability of an audio message to get even the most distracted person to pay attention to the rest of the game is insane, and outside of the occasional shitpost at /tg/ it was very useful for situations like that. The AI serves as the station's watchdog due to Asimov, and being able to ensure every player hears a critical message is important.

**Comment by zjdtmkhzt** [comment]
I really think spam probably won't be an issue.
For one, an AI voice message feels less intrusive to me than some radio song that I have already heard *literally* hundreds of times, because the AI message is likely to be something new and vaguely related to the round.
And also, radio people probably spam those so much because there is little else for them to do. AIs usually have plenty of things to do, so I think they are less likely to overuse these messages.
In addition, it could always have a lengthy cooldown, if an AI seriously overuses it, it could be easily dealt with via a law from a head.

I really think this would be a good and fun tool in an AIs repertoire and I think that if we put it in, soon we will wonder how we ever considered not doing it.

**Comment by Iamgoofball** [comment]
Hell, I'll bet 10 bucks right now that it won't be a problem past the first week, if at all.

**Comment by Iamgoofball** [comment]
i mean shit you don't even need to add a law, you can just go "AI law 2 stop using VOX forever"

**Comment by Gannets** [comment]
This is actual sorcery and I love it.
Kinda wish we had our own voice set rather than the ones trained on existing characters. I don't fundamentally oppose enabling this by default, but maybe we should have it as an admin toggle like VOX is currently, at least for now.

In the future it could be a cool thing to require the crew to assemble a vox machine to enable this ability, maybe even hide different voices behind different modules?

**Comment by MarkNais** [comment]
Spoke with goof;
DNM until 15.ai is brought back online (AWS issues), and other code changes implemented.

**Comment by geneva-breaker** [comment]
Interested in seeing this in game but strongly suggest making it off-by-default for reasons others have detailed, and for my very short patience

**Comment by UrsulaMejor** [comment]
Strongly against the idea of using voice data from people/VAs that didn't give explicit consent for their voices to be used as training data for this project.

**Comment by Iamgoofball** [comment]
As soon as 15 releases his paper I plan to work out something to resolve that.

**Comment by Sovexe** [comment]
Ethics aside I think this is really cool

**Comment by pali6** [comment]
I like this but I don't think it should be on by default for global AI announcements. People already find VOX and Dectalk annoying if admins use it, giving similar power to AI by default would make people mad.

**Comment by Iamgoofball** [comment]
Site's coming back up soon :tm: now that amazon has stopped pranking us.

**Comment by Iamgoofball** [comment]
1 more day, provided amazon doesn't destroy us

**Comment by ZeWaka** [comment]
15.ai is back online, let the ethical debate resume

**Comment by Sovexe** [comment]
I vote merge

**Comment by UrsulaMejor** [comment]
The concerns I brought up still haven't been addressed. it looks like the model doesn't actually need that much data to work, if the web version is to be believed. 60 minutes of training data is not so much as to prohibit using less morally dubious sources

**Comment by Iamgoofball** [comment]
15's currently working on a multi-speaker setup that might let us blast that 60 minute minimum down to an absurdly low amount. 

I'll be adding the requisite toggles and default disables soon.

**Comment by Iamgoofball** [comment]
15.ai'll be back up within the next 2-3 days as per an email I got from 15.

**Comment by geneva-breaker** [comment]
but what if

**Pull request closed without merging**